### ISSUE TITLE
PyTorch Backend Requirements & Design Considerations for Multi-Engine Support

### ISSUE THREAD

**Issue by leocamello** [exo-explore/exo#1347]
## Context

In PR #1284, I've implemented an initial PyTorch backend for NVIDIA GPU support. Based on discussions with @Evanev7, the team is planning a new runner interface, and I'd like to document the requirements and considerations from the PyTorch/NVIDIA
perspective to help inform that design.

## Goals

This issue tracks requirements that a multi-backend runner interface should support:

### 1. Core Inference Requirements

| Requirement | PyTorch Approach | Notes |
|-------------|------------------|-------|
| Model loading | `AutoModelForCausalLM.from_pretrained()` | Supports HuggingFace ecosystem |
| Tokenization | `AutoTokenizer` with `apply_chat_template()` | Model-specific formats |
| Text generation | `model.generate()` with `TextIteratorStreamer` | For streaming output |
| GPU memory | `torch.cuda.memory_summary()` | Different from MLX unified memory |

### 2. Distributed Inference

| Requirement | PyTorch Approach | MLX Difference |
|-------------|------------------|----------------|
| Process group init | `torch.distributed.init_process_group()` | Uses NCCL backend |
| Pipeline parallelism | Layer slicing + send/recv | Similar concept |
| Tensor parallelism | Column/row parallel linear layers | Requires all-reduce per layer |
| **Distributed cancellation** | Broadcast cancel signal across process group | Different latency characteristics |

### 3. Continuous Batching

For reference, here's how PyTorch ecosystems handle this (addressing @Evanev7's concern):

**vLLM Approach:**
- PagedAttention for memory-efficient KV cache
- Scheduler with waiting/running/swapped queues
- Iteration-level scheduling (not request-level)
- Cancellation: Mark sequence as finished, skip in next iteration

**HuggingFace TGI Approach:**
- Rust router + Python inference server
- Token-level batching
- Cancellation: HTTP request abort signals

**Key consideration for exo:**
Distributed cancellation must be coordinated across all nodes. In PyTorch, this typically requires:

```python
# All ranks must agree before removing request from batch
cancel_signal = torch.tensor([1 if cancelled else 0]).cuda()
dist.all_reduce(cancel_signal, op=dist.ReduceOp.MAX)
if cancel_signal.item() > 0:
    # All ranks remove this request
```
### 4. Optimization Features

| Feature | PyTorch Support | Notes |
|---------|-----------------|-------|
| Quantization | bitsandbytes (4/8-bit), GPTQ, AWQ | Different from MLX quantization |
| Flash Attention | `attn_implementation="flash_attention_2"` | Requires compatible GPU |
| KV cache | Built into `generate()` | PagedAttention for efficiency |
| Compilation | `torch.compile()` | Optional but significant speedup |

## Interface Considerations

When designing the new runner interface, consider:

1. **Async generation**: `async def generate()` that yields tokens
2. **Batch awareness**: Support for adding/removing requests dynamically
3. **Cancellation hooks**: Way to signal cancellation that propagates to all nodes
4. **Resource queries**: Memory available, GPU utilization, etc.
5. **Warmup phase**: Some engines need warmup for optimal performance

## Current Implementation

My implementation in PR #1284 provides:
- `Engine` abstract base class with `initialize_distributed_group()`, `load_model_and_tokenizer()`, `warmup_inference()`, `generate()`
- `PytorchEngine` implementation for NVIDIA GPUs
- `MlxEngine` wrapper that encapsulates existing MLX logic
- Pipeline parallelism via layer slicing

Happy to adapt this to whatever interface the team designs. I'll keep PR #1284 rebased and ready.

## References

- [vLLM Paper (PagedAttention)](https://arxiv.org/abs/2309.06180)
- [Orca Paper (Continuous Batching)](https://www.usenix.org/conference/osdi22/presentation/yu)
- [PyTorch Distributed Docs](https://pytorch.org/docs/stable/distributed.html)

**Comment by leocamello** [comment]
## Architecture Clarification (Updated 2026-02-07)

After further reflection on the implementation approach, here are some important clarifications on the Engine abstraction design:

### ✅ Correct Approach: Single Runner, Multiple Engines

The goal is to have **ONE `runner.py`** that uses a well-designed Engine interface:

```
runner.py (engine-agnostic, orchestrates state machine)
  ↓ calls generic engine methods
Engine (abstract interface)
  ├─ MlxEngine (translates to MLX-specific calls)
  ├─ PytorchEngine (translates to PyTorch-specific calls)
  └─ Future: ONNXEngine, TensorRTEngine, TinyGradEngine...
```

**Key principle**: Any functionality that works with MLX must also work with PyTorch (and future engines).

### 🎯 Engine API Design Philosophy

The Engine interface should be **backend-agnostic by design**, not a 1:1 mirror of MLX's API:

| Principle | Example |
|-----------|---------|
| **Intent-based naming** | `cleanup()` not `clear_mlx_cache()` |
| **Generic types** | `GenerationParams` not MLX-specific params |
| **Future-proof** | Design considering ONNX, TensorRT, TinyGrad |
| **Minimal surface** | Only expose what runner.py actually needs |

When MLX adds a feature, we:
1. **Understand the intent** - What problem does this solve?
2. **Design generically** - How would PyTorch/ONNX/TensorRT express this?
3. **Add to Engine interface** - Using clear, backend-agnostic naming
4. **Implement per-engine** - Each engine translates to its native API

### 📋 PytorchEngine Missing Features

Current gaps that need to be addressed for feature parity:

- [ ] **Streaming generation** - Use `TextIteratorStreamer` instead of yielding once at end
- [ ] **Chat template support** - `tokenizer.apply_chat_template()` integration
- [ ] **KV cache management** - Use `past_key_values` in HuggingFace generate
- [ ] **Tool calling** - Parse output for tool call markers
- [ ] **Thinking model support** - Detect and handle `<think>` tags
- [ ] **Proper cleanup** - `torch.cuda.empty_cache()` and distributed cleanup

### 🔄 Rebase Workflow

When upstream changes `runner.py`:
1. Review what MLX functionality was added
2. Design a generic API method (not MLX-specific)
3. Add to Engine interface with clear documentation
4. Implement in **both** MlxEngine and PytorchEngine
5. Update runner.py to use the new engine method

This ensures the Engine abstraction evolves properly and all backends stay in sync.

**Comment by leocamello** [comment]
## Update: PR #1284 Rebased & Review Feedback Addressed (2026-02-18)

PR #1284 has been rebased on latest `main` and all feedback from @AlexCheema's code review has been addressed.

### Key Changes

**Engine-agnostic runner** — `runner.py` now has zero MLX/PyTorch imports. All backend-specific functionality is delegated to the `Engine` ABC:
- `should_cancel(want_to_cancel)` — MlxEngine uses `mx_any()` for distributed coordination; PytorchEngine uses default pass-through
- `initialize_image_model()` / `warmup_image_generator()` / `generate_image()` — image generation support moved to Engine interface
- `cleanup()` — each engine handles its own resource cleanup

**Review fixes:**
- `torch`/`transformers` moved to optional dependencies
- Multi-GPU metrics now aggregate across all GPUs (avg utilization, sum power/VRAM, max temp)
- Token counting uses `tokenizer.encode()` for accuracy
- `cancel_receiver` restored in runner — cancellation fully functional
- Pipeline parallel no-ops documented with clear TODOs
- Style cleanup: `tuple` instead of `Tuple`, `|` instead of `Union`, `Callable` from `collections.abc`

**Validation:** 134 type errors (1 fewer than upstream), all lint checks pass, 198 tests pass.

### Updated Status

| Feature | Status |
|---------|--------|
| Engine abstraction | ✅ Complete — fully engine-agnostic runner |
| Task cancellation | ✅ Complete — works across engines |
| Image generation | ✅ Complete — abstracted via Engine API |
| Optional dependencies | ✅ Complete — torch/transformers are extras |
| Multi-GPU metrics | ✅ Complete — aggregates all GPUs |
| Rebased on main | ✅ Complete — up to date |

PR is ready for another round of review.

**Comment by Evanev7** [comment]
just letting you know a big portion of the batching refactor is now implemented - not with the intent of fully making the runner backend agnostic, but moving in that direction. feedback would be appreciated!
